README.md
# Vire

## What is Vire?
Vire is a simple Vim / Neovim installer and vimrc + plugin or package manager.

Vire makes it simple to install and keep Vim or Neovim up to date on Windows.

Install Vim:

  `vire -i`

Install Neovim:

  `vire -i -n`

This will download and install the binary from Github and extract to the HOME
directory. The directory will also be added to the user path so that starting
`[g]vim.exe` or `nvim[-qt].exe` is easy. It will install either the 32-bit or
64-bit version depending on the architecture of Python since a mismatch will
prevent Python plugins from working.

Installation on Linux is left to the distro package managers.

Vire also makes it super easy to install your `vimrc`. The recommended method is
to maintain the `vimrc` as a gist on Github. [Here's](https://gist.github.com/genotrance/57d327501443f440f522e2c886dadc1b) my vimrc for example.
All Vire needs is the gist ID and it is able to keep the local copy up to date.
All updates can be made on Github and simply running Vire on each machine will
get everything up to date. This strategy makes it simple to keep multiple
machines in sync.

  `vire gistID`

Lastly, Vire does what every other plugin manager does - install every plugin
defined in the vimrc loaded. One advantage of using Vire is that it does not
depend on Git to download plugins and leverages Vim / Neovim's built-in `pack`
method of loading plugins.

Vire only requires Python which is what most modern Vim / Neovim plugins require
and made the most sense to build upon.

## Installation

First download and install a Python distribution if not already present.

Vire can be installed in multiple ways, easiest being from PyPI:

  `pip install vire`

Or as a regular user:

  `pip install vire --user`

If `pip` isn't available or preferred, one of the following methods can be used
to download:

- Clone the latest source:

  `git clone https://github.com/genotrance/vire`

- Download the latest source ZIP:

  `https://github.com/genotrance/vire/archive/master.zip`

Vire along with all dependencies can then be installed to the standard Python
location using:

  `python setup.py install`

Or as a regular user:

  `python setup.py install --user`

Installing Vire as a regular user will require adding `%APPDATA%\Roaming\Python\PythonVER\Scripts`
or `~/.local/bin` to the path.

Vire can also be used without installation with `python vire.py` but will require
all dependencies to be installed manually.

  `pip install pygithub requests` for Python 3.x

  `pip install pygithub requests futures pathlib2` for Python 2.x

After installation, Vire can be run on the command line like an executable.

  `vire -h`

## Typical Scenario

- Download and install Vire

- Install Vim / Neovim if required

  `vire -i [-n]`

- Post vimrc to Github gist if not already

- Backup and remove any existing configuration files and directories so that
  Virc can setup from scratch. NOTE: Vire will not overwrite any existing
  configuration if detected unless told to with `-f`. If forced, the existing
  `vimrc` will be overwritten and the `pack/vire/start` directory will get
  cleaned and repopulated with plugins as specified in the `vimrc` provided. Any
  other directories and files will remain in place and might conflict so it is
  preferred to start with a clean slate.

- If you want to use Vire for all plugins:
  - Edit vimrc to remove all function calls related to plugin managers like Pathogen,
    Vundle, vim-plug, etc.
  - Comment out all lines pointing to plugins. This will be used by Vire to download
    and install instead. For example, for Vundle:

    `Plugin 'tpope/vim-fugitive'` => `" Plugin 'tpope/vim-fugitive'`

  - Or for vim-plug:

    `Plug 'tpope/vim-fugitive'` => `" Plug 'tpope/vim-fugitive'`

  - For other plugins, you  will need to add such `" Plug 'user/project'` manually
    for each plugin that needs to be installed.

- If you prefer to use your existing plugin manager to download your plugins,
  you can use Vire to only install the plugin manager and leave the rest of the
  vimrc as is:

  `" Plug 'VundleVim/Vundle.vim'`
  `" Plug 'junegunn/vim-plug'`
  `" Plug 'Shougo/dein.vim'`

- Run Vire to download and install the vimrc and plugins:

  `vire gistID`

- When you need to update, simply run `vire` and it will check if anything has
  changed. Running with `-i` will also check if the app itself has an update.

## Configuration

Vire maintains a `~/.vire.json` to remember things such as the gist ID, 32-bit
mode, Vim or Neovim mode and other details to detect when an update is available.

No manual editing should be required, the command line parameters should be all
that's required.

## Usage

```
usage: vire [-h] [-b] [-f] [-i] [-n] [-v] [vimrc]

positional arguments:
  vimrc          Gist ID or path to vimrc file

optional arguments:
  -h, --help     show this help message and exit
  -b, --bit32    Force 32-bit install
  -f, --force    Force Vim reinstall
  -i, --install  Install Vim/Neovim
  -n, --neomode  Neovim mode
  -v, --vimmode  Vim mode [default]
```

## Dependencies

Vire only requires Python and the PyGithub module at this time. Setuptools is
also required if using `setup.py`.

Vire is tested on Windows with Python 3.6 using the Miniconda distribution and
on Ubuntu Artful with Python 2.7 and 3.6.

## Feedback

Vire is definitely a work in progress and any feedback or suggestions are welcome.
It is hosted on [GitHub](https://github.com/genotrance/vire) with an MIT license
so issues, forks and PRs are most appreciated.


HISTORY.txt
v0.1.2 - 5/17/2018
- Fix issue #1 - handle global try/except correctly
- Fix issue #2 - add requests as a module dependency
- Fix issue #3 - avoid overwriting existing vimrc on first run

v0.1.1 - 5/16/2018
- Fix pip install issue with setup.py on Python 2.x

v0.1.0 - Initial release
- Add support for installing Vim and Neovim on Windows
- Download and setup vimrc from gist or local copy
- Download all plugins defined in vimrc


vire.py
"Vim / Neovim installer and vimrc + plugin manager"

from __future__ import print_function

import argparse
import ctypes
import datetime
import glob
import io
import json
import os
import platform
import re
import requests
import shutil
import sys
import tempfile
import time
import traceback
import zipfile

try:
    import github
except ImportError:
    print("Requires module pygithub")
    sys.exit()

try:
    import pathlib
except ImportError:
    try:
        import pathlib2 as pathlib
    except ImportErrror:
        print("Requires module pathlib2")
        sys.exit()

if os.name == "nt":
    try:
        import winreg
    except ImportError:
        try:
            import _winreg as winreg
        except ImportError:
            print("Requires module winreg")
            sys.exit()

    try:
        import ctypes.wintypes
    except ImportError:
        print("Requires module ctypes.wintypes")
        sys.exit()

__version__ = "0.1.2"

class State(object):
    config = ""
    force = False
    home = os.getenv("HOME") or ""
    install = False
    packpath = ""
    nvimpath = ".local/share/nvim"
    nvimrcpath = ".config/nvim/init.vim"
    vimpath = ".vim"
    vimrcpath = ".vimrc"
    windows = False

    github = None

Config = {
    "bit64": False,
    "neomode": False,
    "neoversion": "",
    "plugins": {},
    "vimrc": "",
    "vimrcupdated": "",
    "vimversion": ""
}

CConfig = {}
Plugins = []

def parsecli():
    parser = argparse.ArgumentParser()
    parser.add_argument("vimrc", nargs="?", help="Gist ID or path to vimrc file")
    parser.add_argument("-b", "--bit32", action="store_true", help="Force 32-bit install")
    parser.add_argument("-f", "--force", action="store_true", help="Force Vim reinstall")
    parser.add_argument("-i", "--install", action="store_true", help="Install Vim/Neovim")
    parser.add_argument("-n", "--neomode", action="store_true", help="Neovim mode")
    parser.add_argument("-v", "--vimmode", action="store_true", help="Vim mode [default]")

    args = parser.parse_args()
    if Config["bit64"] == True and args.bit32 == True:
        Config["bit64"] = False
    State.force = args.force
    State.install = args.install

    if args.vimrc:
        Config["vimrc"] = args.vimrc

    if args.neomode and args.vimmode:
        print("You really want both Vim and Neovim?")
        sys.exit()

    if args.neomode:
        Config["neomode"] = True
    elif args.vimmode:
        Config["neomode"] = False

def within_rate_limit():
    rl = State.github.get_rate_limit()
    if not rl.rate.remaining:
        print("Github rate limit exceeded, wait until", str(rl.rate.reset), "GMT")
        return False

    return True

def setup():
    global Config
    global CConfig

    if os.name == "nt":
        State.windows = True
        State.home = os.getenv("USERPROFILE")
        State.nvimpath = "AppData/Local/nvim"
        State.nvimrcpath = "AppData/Local/nvim/init.vim"
        State.vimpath = "vimfiles"
        State.vimrcpath = "_vimrc"

    State.nvimpath = os.path.join(State.home, State.nvimpath)
    State.nvimrcpath = os.path.join(State.home, State.nvimrcpath)
    State.vimpath = os.path.join(State.home, State.vimpath)
    State.vimrcpath = os.path.join(State.home, State.vimrcpath)

    State.config = os.path.join(State.home, ".vire.json")

    #if "64" in platform.machine():
    if sys.maxsize != 2147483647:
        Config["bit64"] = True

    if os.path.exists(State.config):
        with open(State.config) as fp:
            CConfig = json.load(fp)
            Config.update(CConfig)

    parsecli()

    if Config["neomode"]:
        State.packpath = os.path.join(State.nvimpath, "pack", "vire", "start")
    else:
        State.packpath = os.path.join(State.vimpath, "pack", "vire", "start")

    pathlib.Path(State.packpath).mkdir(parents=True, exist_ok=True)

    State.github = github.Github()

def extract(zfilename, destination):
    print("- Extracting " + os.path.basename(zfilename))
    with zipfile.ZipFile(zfilename) as zf:
        zf.extractall(destination)

def extract_asset(zfilename):
    if os.path.exists(os.path.join(State.home, "vim")):
        return

    extract(zfilename, State.home)

def download(url, filename):
    print("- Downloading " + url)
    with requests.get(url, stream=True) as resp, open(filename, 'wb') as fp:
        if resp.ok:
            shutil.copyfileobj(io.BytesIO(resp.content), fp)
        else:
            print("Failed download of " + url)
            return False

    return True

def download_asset(asset):
    zfilename = os.path.join(tempfile.gettempdir(), asset.name)
    if os.path.exists(zfilename) and os.stat(zfilename).st_size == asset.size:
        return zfilename

    if not download(asset.browser_download_url, zfilename):
        return ""

    return zfilename

def add_to_path():
    vimapppath = ""
    if not Config["neomode"]:
        for dr in glob.glob(os.path.join(State.home, "vim", "vim*")):
            for fn in glob.glob(os.path.join(dr, "*.exe")):
                if "vim.exe" in fn:
                    vimapppath = dr
                    break
    else:
        for fn in glob.glob(os.path.join(State.home, "Neovim", "bin", "*.exe")):
            if "nvim-qt.exe" in fn:
                vimapppath = os.path.dirname(fn)
                break

    if vimapppath:
        key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, 'Environment', 0, winreg.KEY_ALL_ACCESS)
        gkey = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r'SYSTEM\CurrentControlSet\Control\Session Manager\Environment', 0)
        value = ""
        gvalue = ""
        try:
            gvalue, _ = winreg.QueryValueEx(gkey, "PATH")
            value, _ = winreg.QueryValueEx(key, "PATH")
        except WindowsError:
            pass

        if vimapppath.lower() not in gvalue.lower() and vimapppath.lower() not in value.lower():
            value = value + os.pathsep + vimapppath if value else vimapppath
            winreg.SetValueEx(key, "PATH", 0, winreg.REG_EXPAND_SZ, value)
            print("- Adding Vim to path: " + vimapppath)

            SendMessageTimeout = ctypes.windll.user32.SendMessageTimeoutW
            HWND_BROADCAST = 0xFFFF
            WM_SETTINGCHANGE = 0x1A
            SMTO_ABORTIFHUNG = 0x0002

            try:
                SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE, 0, "Environment",
                    SMTO_ABORTIFHUNG, 5000)
            except ValueError:
                pass

        winreg.CloseKey(gkey)
        winreg.CloseKey(key)

def delete_vim():
    apppath = "vim"
    if Config["neomode"]:
        apppath = "Neovim"
    apppath = os.path.join(State.home, apppath)

    print("- Deleting", apppath)
    if os.path.exists(apppath):
        shutil.rmtree(apppath)

def get_vim():
    if not State.install:
        return

    name = "Vim"
    repo = "vim/vim-win32-installer"
    version = "vimversion"
    if Config["neomode"]:
        name = "Neovim"
        repo = "neovim/neovim"
        version = "neoversion"

    if not State.windows:
        print("Use package manager to install", name)
        return

    latest = State.github.get_repo(repo).get_latest_release()
    if State.force or latest.tag_name > Config[version] or Config["bit64"] != CConfig["bit64"]:
        print("Updating", name+":", Config[version] + " => " + latest.tag_name)
        for asset in latest.get_assets():
            if "zip" in asset.name and "pdb" not in asset.name:
                if (Config["bit64"] and "64" in asset.name) or (not Config["bit64"] and ("x86" in asset.name or "32" in asset.name)):
                    delete_vim()
                    zfilename = download_asset(asset)
                    if zfilename:
                        extract(zfilename, State.home)
                        add_to_path()
                        Config[version] = latest.tag_name
    else:
        print(name, "up to date:", Config[version])

def get_gist(vimrcpath):
    gist = State.github.get_gist(Config["vimrc"])
    if State.force or "vimrcupdated" not in Config or str(gist.updated_at) > Config["vimrcupdated"]:
        for (gfname, gfile) in gist.files.items():
            if "vimrc" in gfname:
                print("Updating", vimrcpath, "with", gfname)
                with open(vimrcpath, "w") as fp:
                    shutil.copyfileobj(io.StringIO(gfile.content), fp)
                Config["vimrcupdated"] = str(gist.updated_at)
                break
    else:
        print(vimrcpath, "up to date")

def get_plugin(reponame):
    global Plugins

    if os.path.splitext(reponame)[1] == ".git":
        reponame = reponame[:-4]
    reponame = reponame.replace("\\", "/")

    plugin = os.path.basename(reponame)
    pluginpath = os.path.join(State.packpath, plugin)

    Plugins.append(plugin)

    if not plugin in Config["plugins"]:
        Config["plugins"][plugin] = {
            "last_checked": time.time(),
            "sha": ""
        }
    elif not State.force and Config["plugins"][plugin]["last_checked"] + 3600 > time.time():
        return

    repo = State.github.get_repo(reponame)
    commits = repo.get_commits()
    latest = commits[0].sha

    Config["plugins"][plugin]["last_checked"] = time.time()

    if latest != Config["plugins"][plugin]["sha"]:
        if os.path.exists(pluginpath):
            print("Updating plugin", reponame)
            shutil.rmtree(pluginpath)
        else:
            print("Loading plugin", reponame)
        zplugin = os.path.join(tempfile.gettempdir(), plugin + "-" + latest + ".zip")
        if not os.path.exists(zplugin):
            if not download("https://github.com/" + reponame + "/archive/" + latest + ".zip", zplugin):
                return
        extract(zplugin, State.packpath)
        for pdir in glob.glob(pluginpath + "-*"):
            os.rename(pdir, pluginpath)

        Config["plugins"][plugin]["sha"] = latest

def get_vimrc():
    if Config["vimrc"] == "":
        print("No vimrc specified")
        return

    dest = State.nvimrcpath if Config["neomode"] else State.vimrcpath
    if Config["vimrcupdated"] == "" and os.path.exists(dest) and not State.force:
        print("vimrc already exists - run with -f to overwrite")
        return

    pathlib.Path(os.path.dirname(dest)).mkdir(parents=True, exist_ok=True)
    if os.path.exists(Config["vimrc"]):
        print("Copying", Config["vimrc"], "to", dest)
        shutil.copyfile(Config["vimrc"], dest)
    else:
        get_gist(dest)

    if os.path.exists(dest):
        with open(dest) as fp:
            for line in fp.readlines():
                match = re.match("\"[ ]+Plug(?:in)?[ ]+'(.*?)'", line)
                if match:
                    get_plugin(match.group(1))

    # Cleanup
    for pdir in glob.glob(os.path.join(State.packpath, "*")):
        if os.path.basename(pdir) not in Plugins:
            print("Removing", pdir)
            shutil.rmtree(pdir)

    if "plugins" in Config:
        for plugin in list(Config["plugins"]):
            if not os.path.exists(os.path.join(State.packpath, plugin)):
                del Config["plugins"][plugin]

def save():
    with open(State.config, "w") as fp:
        json.dump(Config, fp)

def main():
    try:
        setup()
        get_vim()
        get_vimrc()
    except github.GithubException as e:
        if e.status == 403:
            within_rate_limit()
        elif e.status == 404:
            print("Gist ID not found")
        else:
            print(e)
    except SystemExit:
        pass
    except:
        traceback.print_exc(file=sys.stdout)

    save()

if __name__ == "__main__":
    main()
